cpp/src/text/edit_distance.cu

```diff
@@ -101,34 +100,27 @@ struct edit_distance_levenshtein_algorithm {
     auto d_str =
       d_strings.is_null(idx) ? cudf::string_view{} : d_strings.element<cudf::string_view>(idx);
     auto d_tgt = [&] __device__ {  // d_targets is also allowed to have only one entry
-      if (d_targets.is_null(idx)) { return cudf::string_view{}; }
+      if (d_targets.size() > 1 && d_targets.is_null(idx)) { return cudf::string_view{}; }
       return d_targets.size() == 1 ? d_targets.element<cudf::string_view>(0)
                                    : d_targets.element<cudf::string_view>(idx);
     }();
     d_results[idx] = compute_distance(d_str, d_tgt, d_buffer + d_offsets[idx]);
   }
 };
 
-struct edit_distance_matrix_levenshtein_algorithm {
-  cudf::column_device_view d_strings;  // computing these against itself
-  cudf::size_type* d_buffer;           // compute buffer for each string
-  std::ptrdiff_t const* d_offsets;     // locate sub-buffer for each string
-  cudf::size_type* d_results;          // edit distance values
+struct calculate_compute_buffer_fn {
+  cudf::column_device_view d_strings;
+  cudf::column_device_view d_targets;
 
-  __device__ void operator()(cudf::size_type idx) const
+  __device__ std::ptrdiff_t operator()(cudf::size_type idx) const
   {
-    auto const strings_count = d_strings.size();
-    auto const row           = idx / strings_count;
-    auto const col           = idx % strings_count;
-    if (row > col) return;  // bottom half is computed with the top half of matrix
-    cudf::string_view d_str1 =
-      d_strings.is_null(row) ? cudf::string_view{} : d_strings.element<cudf::string_view>(row);
-    cudf::string_view d_str2 =
-      d_strings.is_null(col) ? cudf::string_view{} : d_strings.element<cudf::string_view>(col);
-    auto work_buffer    = d_buffer + d_offsets[idx - ((row + 1) * (row + 2)) / 2];
-    auto const distance = (row == col) ? 0 : compute_distance(d_str1, d_str2, work_buffer);
-    d_results[idx]      = distance;                   // top half of matrix
-    d_results[col * strings_count + row] = distance;  // bottom half of matrix
+    if (d_strings.is_null(idx)) { return 0; }
+    if (d_targets.size() > 1 && d_targets.is_null(idx)) { return 0; }
+    auto d_str = d_strings.element<cudf::string_view>(idx);
+    auto d_tgt = d_targets.size() == 1 ? d_targets.element<cudf::string_view>(0)
+                                       : d_targets.element<cudf::string_view>(idx);
+    // just need 2 integers for each character of the shorter string
+    return (cuda::std::min(d_str.length(), d_tgt.length()) + 1) * 2;
```

Comment:
Will this overflow by any chance? `d_str.length()` may be `INT_MAX` then `+1` will overflow before it is promoted to the returned type.

Reply:
Fixed